From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Andrew den Exter <andrew.den.exter@jolla.com>
Date: Thu, 11 Nov 2021 06:51:59 +0000
Subject: [PATCH 60/67] Disallow page zooming for a fixed meta viewport

Respect fixed meta viewport scales and user-scalable=no when deciding whether
APZ may zoom a document. Continue to allow zooming in responsive-design mode
and when EmbedLite forces the top-level browsing context into desktop mode.

JB#56149 OMP#JOLLA-49
JB#56767 OMP#JOLLA-609
---
 layout/base/nsLayoutUtils.cpp | 19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

diff --git a/layout/base/nsLayoutUtils.cpp b/layout/base/nsLayoutUtils.cpp
index aa6ef4435df0..1b2488cb548f 100644
--- a/layout/base/nsLayoutUtils.cpp
+++ b/layout/base/nsLayoutUtils.cpp
@@ -79,6 +79,7 @@
 #include "mozilla/ViewportUtils.h"
 #include "mozilla/WheelHandlingHelper.h"  // for WheelHandlingUtils
 #include "mozilla/dom/AnonymousContent.h"
+#include "mozilla/dom/BrowsingContext.h"
 #include "mozilla/dom/BrowserChild.h"
 #include "mozilla/dom/CanvasUtils.h"
 #include "mozilla/dom/CharacterDataBuffer.h"
@@ -696,6 +697,20 @@ bool nsLayoutUtils::AsyncPanZoomEnabled(const nsIFrame* aFrame) {
   return widget->AsyncPanZoomEnabled();
 }
 
+static bool DocumentDisallowsZooming(const Document* aDocument) {
+  const ViewportMetaData metaData = aDocument->GetViewportMetaData();
+  return (!metaData.mMinimumScale.IsEmpty() &&
+          metaData.mMinimumScale == metaData.mMaximumScale) ||
+         metaData.mUserScalable.EqualsLiteral("0") ||
+         metaData.mUserScalable.EqualsLiteral("no") ||
+         metaData.mUserScalable.EqualsLiteral("false");
+}
+
+static bool IsDesktopView(const Document* aDocument) {
+  BrowsingContext* bc = aDocument->GetBrowsingContext();
+  return bc && bc->Top()->ForceDesktopViewport();
+}
+
 bool nsLayoutUtils::AllowZoomingForDocument(
     const mozilla::dom::Document* aDocument) {
   if (aDocument->GetPresShell() &&
@@ -705,7 +720,9 @@ bool nsLayoutUtils::AllowZoomingForDocument(
   // True if we allow zooming for all documents on this platform, or if we are
   // in RDM.
   BrowsingContext* bc = aDocument->GetBrowsingContext();
-  return StaticPrefs::apz_allow_zooming() || (bc && bc->InRDMPane());
+  return (StaticPrefs::apz_allow_zooming() &&
+          !DocumentDisallowsZooming(aDocument)) ||
+         (bc && bc->InRDMPane()) || IsDesktopView(aDocument);
 }
 
 static bool HasVisibleAnonymousContents(Document* aDoc) {
